### ISSUE TITLE
[css-fonts-4] Managing the scaling of font-size within MathML rendering

### ISSUE THREAD

**Issue by bkardell** [w3c/csswg-drafts#5389]
(Part of #5384 - MathML Core related CSS)

Note: We're very interested in CSSWG feedback on this - it has already changed once based on feedback and tests updated, etc - and there has been some additional feedback that perhsaps this should be a single value instead which indicates that perhaps it should consult another (new property).  We're ivery nterested to nail this one down better.

---

Script levels in MathML are about managing the font-size in subexpressions (like scripts and fractions).  [This example](https://mathml-refresh.github.io/mathml-core/#font-size-scriptlevel-example) illustrates and describes some of the concepts/implications.  The method for doing this is not quite as simple as inheriting, so the MathML-Core spec introduces a new `scriptlevel()` function (See example in the [MathML Core spec](https://mathml-refresh.github.io/mathml-core/#the-math-script-level-property) )
 
Depending on context (exponents, fractions, etc) and particulars of rendering style (compact/inline), the font-size of of parts of expressions scale down at different and fairly well established rates - for example, the scaling factor adjustment of a superscript is not the same as that of a root's prescript.  The example below shows several contextual scaling variations.

![rendering of An exponent with an exponent, a root with a prescript and a fraction containing another fraction](https://mathml-refresh.github.io/mathml-core/examples/example-font-size-scriptlevel.png)

The common rules derive from TexBook, but they are subtle and deserve to be exposed for tweaking and extension by authors.  To this end, MathML adds a new `scriptlevel` function for CSS Fonts Module Level 4 which allows authors to express that the value should be automatically derived (default), or explicitly set to a specific level or that it should be the current level + some integer with rules defined in [section 4.5 of the spec](https://mathml-refresh.github.io/mathml-core/#the-math-script-level-property)

Tests in https://github.com/web-platform-tests/wpt/tree/master/css/css-fonts/math-script-level-and-math-style

**Comment by faceless2** [comment]
We'd implemented the original `math-script-level` property while the core spec was evolving. For what it's worth, this proposal looks a lot better to me. An awful lot better.

Only one suggestion really - would it be an idea to reset the internal script-level to 0 if the font-size is specified as anything other than a `scriptlevel()` function?

I ask because, as specified, "the internal scriptlevel is 0 on the root element" - that's the root element of the document, not of the math equation?

If so, then my issue is that I can't reset the internal script level _and_ change the font-size at the same time. And I don't see anything that limits the use of this to MathML element only. 

```html
<html style="font-size: 16px"> <!-- root element. script-level starts at 0 -->
 <div style="font-size: scriptlevel(1)">
  This text is 11.36px
  <div style="font-size: 16px">
   This text is 16px again, because the font-size was specified as a length, not a `scriptlevel()` function.
   <math>
    <mtext>This is still 16px</mtext>
    <munderover> <!-- user-agent stylesheet increments script-level of 2nd and 3rd children -->
     <mo>&#x220F;</mo>
     <mi>a</mi> <!-- I'd expected 11.36px, but it's actually 8px because scriptlevel is 2, not 1 -->
     <mi>b</mi>
    </munderover>
   </math>
  </div>
 </div>
</html>
```
I could get around this by adding `font-size: scriptlevel(0)` to the `<math>` element. But if I also wanted to set the `font-size` on the math element to something constant like 20px, I couldn't do both.

It's obviously not very likely, but having a property like internal-script-level that isn't visible to the user, that inherits all the way through the node hierarchy and that can't be reset without also resetting another property feels a little sketchy. The above would be a real pig for an author to debug.

I don't know how often a fixed font-size is used for internal elements in MathML - in our batch of testcases here, only once, within an `<mtable>`. Resetting the script-level to zero at that point looks reasonable to me.

In short, my only suggestion would be to

1. Reset the internal-script-level to 0 if font-size is set to anything other than a `scriptlevel()` function
2. Add `math { font-size: scriptlevel(0) }` to the MathML user-agent stylesheet.

The combination shuold ensure that a `<math>` element always starts with an internal-script-level of 0, whether the user specifies a font-size on the math element or not.

Alternatively, specify that the internal script-level is reset to 0 on a `<math>` element by some internal means unrelated to the `font-size` property.

**Comment by fantasai** [comment]
Fwiw, I think my original concern was that the previous iteration of this feature--which used a separate property as the counter--depended on concepts that don't exist in CSS, like whether `font-size` has a specified value. (It always has a specified value.) Integration with font-size has to be defined in a way that can actually work; however, doesn't mean necessarily that the entire feature needs to be encapsulated into the font-size property itself.

We could keep the depth counter as its own property, something like
```
math-depth: auto | add || <integer>
inherited: yes
computed value: integer
```

Then for 
> If the specified value font-size is of the form scriptlevel(...) then the computed value of font-size is obtained by multiplying the inherited value of font-size by a nonzero scale factor calculated by the following procedure ...
you could instead have something like

- Option A: The used value of `font-size` is obtained by multiplying the computed `font-size` by a nonzero scale factor calculated by the following procedure [ insert MathML algorithm, with A=0 ] ...
- Option B: If [ `math-depth` is nonzero | the specified `font-size` is `math` ] , the computed value of `font-size` is obtained by multiplying the inherited `font-size` by a nonzero scale factor calculated by the following procedure ...
- Option C: If `math-depth` is nonzero, the computed value of `font-size` is obtained by multiplying what would have been the computed `font-size` by a nonzero scale factor calculated by the following procedure ...

Here's the side-effects of various options:
- Option A: Here, the scale factor is inherited independently of the inherited font size, and they're combined at used-value time. This means that setting `font-size` to a value doesn't necessarily give you that value: it gives you that value multiplied by the effective scale factor tracked by `math-depth`. It means that font-relative length units are relative to the base `font-size` rather than to the effective one; you may or may not want this effect.
- Option B: The `math-depth` calculation overrides the specified `font-size`. In the first sub-option (“if `math-depth` is nonzero”) this override is implied, and whatever the author specified on `font-size` is ignored. In the second (“if `font-size` is `math`”), this interaction is explicit; the author can't specify a conflicting value on `font-size`, since they have to specify the `math` keyword to explicitly defer the font-size calculation to `math-depth`.
- Option C: In this case, the math scaling operates on the “raw” computed font-size: in most cases, this will be the inherited `font-size` (as in the previous option), but if the author specifies a `font-size` on the element, then that value will be used as the basis for scaling instead of the inherited value.

Option D: If you really want a declaration like `font-size: <length>` to reset the math-depth counter so that you get that absolute size, though, then `font-size` is essentially a shorthand for a depth counter property and a length tracker property -- your current spec is effectively doing this, but without real longhands for this `font-size` “shorthand”. I think it's a bit weird to have `font-size` (which is such a basic property) incorporate this separate concept of a depth counter into its computed value, but maybe it's okay... would definitely want to get the rest of the CSSWG's opinion on that, though.

I'll note that we're running into a similar need for a font-size scaling factor for drop caps, see https://github.com/w3c/csswg-drafts/issues/4988 ; it would be interesting to note what the similarities and differences are wrt expected behavior. (For example, in the drop caps case, we suspect that font-relative lengths should be unaffected by `initial-letter`’s scaling effects.)

**Comment by fred-wang** [comment]
Thanks @fantasai !

MathML3 says that an explicit font-size overrides the effect of scriptlevel change, so that excludes option A, B-1 and C.

Option B-2 looks more or less what MathML Core was trying to achieve in the orignal proposal. To make it work, MathML Core would only need to ensure font-size is set to math at least for the default UA stylesheet inheritance and when the scriptlevel attribute is used ; or maybe even on all MathML elements but the `<math>` root.

Additionally, using this explicit `math` keyword will allow authors to modify the `math-depth` without a font-size change as a side effect. In particular, the MathML UA stylesheet could do:

    math {
      font-size: inherited;
      math-depth: 0;
    }

to ensure the math-depth is 0 on the `<math>` root and so absolute values of scriptlevel are correctly interpreted.

About option D, I don't think that's necessary for the MathML use case. The reason why I tried to incorporate it into font-size was to simplify the proposal to make it more easily acceptable to the CSSWG but if people are happy with a separate math-depth counter, I wouldn't add more complexity.

**Comment by fantasai** [comment]
Probably 'auto' should have a different, less-confusing name. Maybe 'auto-add'? 'auto-compact'?

**Comment by fred-wang** [comment]
The proposal has been updated to use option B-2:
https://mathml-refresh.github.io/mathml-core/#the-math-script-level-property

**Comment by svgeesus** [comment]
Is the new `math-depth` property being added to MathML, to CSS Fonts 4, or someplace else?

**Comment by svgeesus** [comment]
@fred-wang @bkardell `math-depth` is going where?

**Comment by svgeesus** [comment]
Ping @fred-wang which specification gets the proposed `math-depth` property?

**Comment by fred-wang** [comment]
@svgeesus: Sorry for missing your ping, I've been very busy with non-spec work in the past months.

That's a good question and it has also been confusing for me because I initially thought the idea was to add the new properties to the relevant CSS specs but @fantasai suggested to keep stuff in the MathML Core spec when I sent the PR https://github.com/w3c/csswg-drafts/pull/5536#issuecomment-697114884

Anyway, I don't have strong preference on my side, I'm fine with "patching" the CSS spec in MathML Core for now. Since you already submitted https://github.com/w3c/csswg-drafts/commit/09efeb0f1948f5aec4afc3672c1ab0c264abab36 I can simplify that part in the MathML Core spec, though. Thanks!
